trieste/acquisition/function.py

```diff
@@ -676,3 +676,243 @@ def batch_ei(at: TensorType) -> TensorType:
+    r"""
+    Builder of the acquisition function maker for greedily collecting batches by local
+    penalization.  The resulting :const:`AcquisitionFunctionMaker` takes in a set of pending
+    points and returns the expected improvment acquisiton function penalized around those points.
+
+    Local penalization allows us to perform batch Bayesian optimization with a standard (non-batch)
+    acqusition function. By iteratively building a batch of points by maximizing an acquisition
+    function that is clipped around the values correponding to locations close to the already
+    chosen (pending) points, local penalization provides diverse batches of canidadate points.
+
+    Local penalization is applied to the acquisition function multiplicatively. However, to
+    improve numerical stability, we perfom additive penalization in a log space.
+
+    For our penalization function, we implement the soft penalization strategy
+    of :cite:`Gonzalez:2016`, where an estimate of the objective function's Lipschitz
+    constant is used to control the size of penalization.
+
+    The Lipschitz constant and additional penalization parameters are estimated once
+    when first preparing the acquisition function with no pending points. These estaimtes
+    are reused for all subsequent function calls.
+    """
+
+    def __init__(self, search_space: SearchSpace, grid_size: int = 500):
+        """
+        :param search_space: The global search space over which the optimisation is defined.
+        :param grid_size: Size of the grid over which the Lipschitz constant is estimated.
+            We recommend scaling this with search space dimension.
+        """
+        self._search_space = search_space
+        if grid_size <= 0:
+            raise ValueError(f"grid_size must be positive, got {grid_size}")
+        self._grid_size = grid_size
+
+        self._lipschitz_constant = None
+        self._eta = None
+
+    def prepare_acquisition_function(
+        self,
+        dataset: Dataset,
+        model: ProbabilisticModel,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data from the observer.
+        :param model: The model over the specified ``dataset``.
+        :param pending_points: The points we penalize with respect to.
+        :return: The (log) expected improvement penalized with respect to the pending points.
+        :raise ValueError: if the first call does not have pending_points=None.
+        """
+        if len(dataset.query_points) == 0:
+            raise ValueError("Dataset must be populated.")
+
+        if (
+            pending_points is None
+        ):  # only compute penalization parameters once per optimization step
+            samples = self._search_space.sample(num_samples=self._grid_size)
+            samples = tf.concat([dataset.query_points, samples], 0)
+
+            def get_lipschitz_estimate(
```

Comment:
slightly confusing that this is here and not part of the `PenalizationFunction`. Maybe it's not possible, but this is what would makes sense (e.g. we could use a different penalty that doesn't need the Lischitz constant)

Reply:
I guess it could go in the penalization function. I'll see what that looks like